src/index.js

```diff
@@ -542,13 +548,40 @@ class Libp2p extends EventEmitter {
  * @returns {Libp2p}
  */
 Libp2p.create = async function create (options = {}) {
-  if (options.peerId) {
-    return new Libp2p(options)
+  let peerId = options.peerId
+
+  if (!peerId) {
+    peerId = await PeerId.create()
+
+    options.peerId = peerId
   }
 
-  const peerId = await PeerId.create()
+  const keychainOptions = options.keychain || {}
+
+  if (keychainOptions.pass) {
+    log('creating keychain')
+
+    const datastore = keychainOptions.datastore || new MemoryDatastore()
+    const keychainOpts = Keychain.generateOptions()
+
+    const keychain = new Keychain(datastore, {
+      passPhrase: keychainOptions.pass,
+      ...keychainOpts,
+      ...keychainOptions
+    })
+
+    log('keychain constructed')
+
+    // Import the private key as 'self', if needed.
+    try {
+      await keychain.findByName('self')
+    } catch (err) {
+      await keychain.importPeer('self', peerId)
+    }
+
+    options._keychain = keychain
```

Comment:
Having keychain and _keychain on options is confusing. Why not move the keychain creation in to the libp2p constructor? We can create libp2p then do the keychain await stuff before returning the libp2p instance.